lib/Dialect/FIRRTL/Transforms/ModuleSummary.cpp

```diff
@@ -0,0 +1,99 @@
+//===- ModuleSummaryt.cpp ---------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "mlir/IR/Threading.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+struct ModuleSummaryPass : public ModuleSummaryBase<ModuleSummaryPass> {
+  struct KeyTy {
+    SmallVector<size_t> portSizes;
+    size_t opcount;
+    bool operator==(const KeyTy &rhs) const {
+      return portSizes == rhs.portSizes && opcount == rhs.opcount;
+    }
+  };
+
+  size_t countOps(FModuleOp mod) {
+    size_t retval = 0;
+    mod.walk([&](Operation *op) { retval += 1; });
+    return retval;
+  }
+
+  SmallVector<size_t> protSig(FModuleOp mod) {
+    SmallVector<size_t> ports;
+    for (auto p : mod.getPortTypes())
+      ports.push_back(cast<FIRRTLBaseType>(cast<TypeAttr>(p).getValue())
```

Comment:
Ports could have PropertyType or RefType s probably `cast<FIRRTLBaseType>` is not safe ?

Reply:
Not when I was running it, but where I want to, yes.